### PR TITLE
Add support for large sampler arrays on Vulkan

### PR DESCRIPTION
**Pull request by gdkchan**
This is yet another take on the bindless textures problem. This change aims to solve a specific problem: Games that use arrays of samplers on the shader. We already have *some* support for this, but it currently uses a hardcoded array length of 4. This does not work for games that uses more than 4 textures in the array. The pattern recognition for the indexed access is also currently rather limited, which stops it from working on some games (like Penny's Big Breakaway).

This change allows an arbitrary array length. The length is now defined based on the size of the constant buffer with the handles. There has been a few changes on the code to allow array bindings to be updated more efficiently, including new backend interfaces (`ITextureArrat` and `IImageArray`) to prevent redundant update of the array bindings, and a more efficient way to change them.

For OpenGL, nothing should change right now. It has a very low limit on the number of samplers per shader stage (32), which makes this approach unfeasible, however Vulkan has no such limits.

This change fixes graphical issues on a few games when using Vulkan API.

Hogwarts Legacy.
Before:
image
After:
image
(Still has glitches).
Penny's Big Breakaway.
Before:
image
After:
image
(Seems to be still unplayable due to other issues).
The Legend of Zelda: Skyward Swod.
Before:
image
After:
image

Other games using sampler arrays but where I don't expect any change (still needs to be tested for regressions):
- Luigi's Mansion 3.
- Mario Tennis Aces.

Testing is welcome.

### CODE REVIEW

**Comment by zombie167** [comment]
This PR has fixed major graphic glitches,such as screen flickering and frequent crashes.Thank you very much !

**Comment by gdkchan** [comment]
Ready for review.

**Review by riperiperi — APPROVED** [review]
Looks good, tested a bunch of games and they seem fine, though I don't have too many that are affected by bindless stuff.

**Comment by StevensND** [comment]
1 more approval ... 🙏

File: src/Ryujinx.Graphics.Vulkan/ImageArray.cs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        private HashSet<TextureStorage> _storages;
+
+        private int _cachedCommandBufferIndex;
+        private int _cachedSubmissionCount;
+
+        private readonly bool _isBuffer;
+
+        public bool Bound;
+
+        public bool IsDirty => _storages == null;
```
[the remark this answers is not in the document]
Comment (gdkchan) [reply src/Ryujinx.Graphics.Vulkan/ImageArray.cs:34]: This is unused, I forgot to remove it.

File: src/Ryujinx.Graphics.Gpu/Image/TexturePool.cs [hunk 2]
```diff
@@ -432,7 +432,7 @@ private static TextureInfo GetInfo(in TextureDescriptor descriptor, out int laye
             {
                 if (gpuVa != 0 && format != 0)
                 {
-                    Logger.Error?.Print(LogClass.Gpu, $"Invalid texture format 0x{format:X} (sRGB: {srgb}).");
+                    // Logger.Error?.Print(LogClass.Gpu, $"Invalid texture format 0x{format:X} (sRGB: {srgb}).");
```
[the remark this answers is not in the document]
Comment (gdkchan) [reply src/Ryujinx.Graphics.Gpu/Image/TexturePool.cs]: I enabled it again. It was spamming on Hogwarts Legacy. Still does, but it's a little better thanks to the array cache. We can remove it or change to Debug log level, but I think that decision can be done separately.

**Comment by ParanoidHitman** [comment]
Game still has freezes when using spells and loading textures on mac/windows, texture are good, but lights flicker and sometimes goes wrong. Still need optimization for Hogwarts Legacy.

**Comment by jhorv** [comment]
> Game still has freezes when using spells and loading textures on mac/windows, texture are good, but lights flicker and sometimes goes wrong. Still need optimization for Hogwarts Legacy.

@ParanoidHitman I'm curious if you're able to try the version in PR #6623 and see if there's a performance improvement?

**Pull request merged**